Make connection and pool creation fully async (JAVA-692).

olim7t opened the pull request:
Connection and HostConnectionPool constructors don't block anymore, both classes have a new initFuture that clients must wait on.

Hint for reviewers: start in Connection and make your way up.

olim7t committed:
Make connection and pool creation fully async.

olim7t commented:
Early preview, I've not tested it extensively but a few randomly picked tests pass.
Nor have I benchmarked yet.

tolbertam commented:
This looks really good and was much more straightforward to follow then I anticipated! (as usually making code more async makes it harder to understand, but that was not the case here)

Some initial testing reveals a great performance improvement, even with just 3 hosts w/ 2 core connections using password authentication.  The mean session.connect() time is 104.32ms with this branch vs. 167.42ms on 2.0.9.2.   With 8 core connections its 366.69ms vs. 690.13ms.  I imagine the benefit being more pronounced with more hosts then the executor pool size.

olim7t committed:
Check cluster name after successful auth.

olim7t commented:
I've run the integration suite locally and I get the following errors:

```
Failed tests: 
  AuthenticationTest.testConnectionAttemptWithIncorrectCredentialsIsRefused » Test
  AuthenticationTest.testConnectionAttemptWithoutCredentialsIsRefused » Test 
Ex...
  ExceptionsTest.readTimeoutException:195 » InvalidQuery Unmatched column names/...
  ExceptionsTest.unavailableException:311 » InvalidQuery Unmatched column names/...
  ExceptionsTest.writeTimeoutException:364 » InvalidQuery Unmatched column names...
  NettyOptionsTest.should_invoke_netty_options_hooks:54 NoClassDefFound com/data...
  NettyOptionsTest.should_invoke_netty_options_hooks:54 NoClassDefFound com/data...
  RecommissionedNodeTest.should_ignore_node_that_does_not_support_protocol_version_on_session_init:121 expected:<[DOWN]> but was:<[ADDED]>
  SessionTest.sessionMemoryLeakTest:184 NullPointer
  SpeculativeExecutionTest.beforeMethod:37 NoSuchMethod com.datastax.driver.core...
  SSLAuthenticatedEncryptionTest>SSLTestBase.beforeClass:36 » Runtime
  SSLEncryptionTest>SSLTestBase.beforeClass:36 » Runtime
  StateListenerTest.listenerTest:35 » Runtime
  StatementWrapperTest>CCMBridge$PerClassSingleNodeCluster.beforeClass:471->CCMBridge$PerClassSingleNodeCluster.initKeyspace:511->configure:26 NoSuchMethod
```

I'll check which are related to the connection changes.

tolbertam commented:
I think the SessionTest#sessionMemoryLeakTest breaks the ones that get executed after, fixed that issue on 2.0 in 7bd8d43b687c94fc790ea428fa90e68fd1370685.

tolbertam commented:
It looks like the `AuthenticationTest` cases are failing because `Connection#launderAsyncInitException` doesn't take into account `AuthenticationException`.  Previously this would not need to be handled since `AuthenticationException` is a `DriverException` which is a `RuntimeException` so it didn't need to be explicitly thrown.  Maybe we should add to `launderAsyncInitException` to check for `RuntimeException` instances and throw them as is to be consistent with the previous implementation?

I rebased on 2.0 and it fixed all the other errors (were caused by `SessionTest#sessionMemoryLeak` test not properly closing a ccm instance).

I added some more performance numbers to the jira issue ([JAVA-692](https://datastax-oss.atlassian.net/browse/JAVA-692)).  I still need to do some more resiliency testing but I should complete that tomorrow.

olim7t committed:
Handle DriverExceptions during connection init.

olim7t commented:
I fixed the code to handle `DriverException`s correctly.
A warning with the following exception occasionally pops up during tests, but it will be fixed once we upgrade to Netty 4.0.27:

```
io.netty.channel.ChannelException: java.net.SocketException: Socket is closed
```

olim7t commented:
Rebased on 2.0 to use new Netty version.

tolbertam committed:
Add license to MoreFutures.java.

tolbertam commented:
Added license to MoreFutures to get the build passing.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -86,38 +92,59 @@ public void run() {
+                @Override
+                public void onSuccess(List<PooledConnection> l) {
+                    connections.addAll(l);
+                    open.set(l.size());
```
One small issue here is that there is a small window where the Pool future is marked as complete, but open is not set yet.   This could cause issues in the various places we check open when spawning and trashing connections.  I've seen this a few times in SessionTest#sessionMemoryLeak test, but it's very hard to reproduce.  I think we need another intermediate future that gets set after open.set completes in onSuccess that initFuture transforms off of.

tolbertam replied:
Now that I think about it a better option would be to change initFuture to a SettableFuture and then set it after open.set(l.size()).  This avoids the need to do a Futures.transform below.

tolbertam replied:
Also need to update onFailure to setException on initFuture.   I'll commit something for this and it can be reverted if there' s a better way to do this.

tolbertam replied:
Updated, hopefully this looks ok, not even completely sure it is necessary.

tolbertam committed:
Don't complete HostConnectionPool#initFuture until open is updated.

tolbertam committed:
Fix memory leak caused by Connection#flusherLookup.

Discovered while running ClusterStressTest that there was a lot of
heap pressure as more Cluster instances were created.  Upon further
investigation found that Connection#flusherLookup was the culprit as it
is not cleaned up as Clusters are closed.  To solve this, made
flusherLookup a concurrent map with weak keys and made Flusher#eventLoop
a weak reference.  This way as the EventLoopGroups become closed their
associated EventLoops will be marked as weak references and eventually
cleaned up by the garbage collector.

tolbertam committed:
Use a higher port range for CCMBridge since it interferes with YourKit.

tolbertam committed:
Backport Stress tests from 2.1 and enhance Socket cleanup validation.

Improved tests that validate leakiness of Sessions and Clusters to also
ensure that all created SocketChannels get closed appropriately.

Created SocketChannelMonitor which utilizes NettyOptions to observe
created SocketChannels and track them.

tolbertam commented:
Pushed a few commits on this branch.  
1. I found a small bug which may have little or no implication in that a `HostConnectionPool` future would complete before 'open'.
2. Identified a memory leak caused by Connection.flushLookup that is only apparent when creating and closing many Cluster objects.
3. Changed CCMBridge to use ports on 11000+ instead of 10000+ (was interfering with YourKit agent)
4. Backported ClusterStress and SessionStressTest from 2.1 as they seemed like they would provide value for this particular issue (helped me identify 1 and 2).

I also have another pending test I'm working on that stresses read and connection timeouts to force pools needing to be closed and recreated.  I plan on pushing that tomorrow.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -665,7 +754,7 @@ public void shutdown() {
 
     private static final class Flusher implements Runnable {
-        final EventLoop eventLoop;
+        final WeakReference<EventLoop> eventLoopRef;
```
Another alternative to this would be to make `flusherLookup` also use `weakValues`.  I tested with this and it appeared to work exactly the same, but I think I like using `WeakReference` directly instead.  Any opinion or ideas to handle this better?

olim7t replied:
Yes, the only reason Flusher keeps a reference to the event loop is to reschedule itself, which makes the whole thing kind of weird (if `run()` is executing, the event loop has obviously not been GC'ed). But I agree that using `WeakReference` here feels slightly better.

On a side note, since this code was copied from Cassandra, couldn't it cause a leak there as well? Event loops also become stale as clients come and go. It's probably less severe, but worth reporting.

tolbertam commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -153,119 +155,165 @@ private static String extractMessage(Throwable t) {
+        if (expected == null)
+            return MoreFutures.VOID_SUCCESS;
+
+        DefaultResultSetFuture clusterNameFuture = new DefaultResultSetFuture(null,new Requests.Query("select cluster_name from system.local"));
```
Ran into a small bug while running a timeout stress test:

```
Exception in thread "cluster1-worker-1" java.lang.NullPointerException
    at com.datastax.driver.core.DefaultResultSetFuture.cancel(DefaultResultSetFuture.java:252)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:827)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:828)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:828)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:837)
    at com.google.common.util.concurrent.Futures$ChainingListenableFuture.cancel(Futures.java:827)
    at com.datastax.driver.core.HostConnectionPool.forceClose(HostConnectionPool.java:135)
    at com.datastax.driver.core.HostConnectionPool.access$400(HostConnectionPool.java:45)
    at com.datastax.driver.core.HostConnectionPool$2.onFailure(HostConnectionPool.java:122)
    at com.google.common.util.concurrent.Futures$4.run(Futures.java:1140)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:695)
```

I think this is because clusterNameFuture doesn't have a handler, so when a connection gets closed while being constructed and was in the process of checking cluster name this could happen.

I'm unsure of the implications, but since forceClose throws an exception as soon as it hits this exception, other connections may not be cancelled/scheduled for forceClose.  I think doing a simple null check on the handler may fix this, but testing to make sure that doesn't introduce any possible problems since handler is set after construction of DefaultResultSetFuture.

olim7t replied:
I agree with your analysis and fix. The potential NPE was already there before, but creating and closing the connections in parallel makes it much more likely to happen.

tolbertam committed:
Fix issue where NPE is thrown when cancelling future for Connection.

If a Connection initialization future is cancelled while checking the
Cluster name, an NPE may be thrown as the handler associated with the
future is null.  This change fixes the issue by doing a null check in
cancel.

tolbertam committed:
Fix connection leak in borrowConnection when initing/closing.

Discovered that when a pool is replaced, borrowConnection may be called
on it before the initFuture has completed and the connections list is
populated.  This caused more connections to be created and associated
with the HostConnectionPool then there should have been.

Also check for 'isClosing' in borrowConnection instead of isClosed since
discardAvailableConnections() is called before the future is set, which
may cause borrowConnection() to create more connections while we are
closing.  This may not be an actual issue, but it seemed safe to do
this.

tolbertam committed:
Add TimeoutStressTest for exercising Pool timeout and replace logic.

tolbertam commented:
Fixed two issues I found when writing TimeoutStress test:
1. The issue described above where a connection init future is cancelled while checking cluster name is fixed by null checking handler in DefaultResultSetFuture.
2. Check for whether or not the Pool is initialized in borrowConnection and throw an Exception if it isn't.  I'm thinking this might not be a complete solution since it throws a ConnectionException, which increments the errors metric which we may not want. Another alternative is to check initFuture in RequestHandler#query and not call borrowConnection if it isn't inited.  This caused connections to be created when they shouldn't have, and these connections seemed to leak (can share heap dump if you like) and not go away when the Session was closed.

tolbertam commented:
I believe I've found another issue where the driver gives up on a Host when an exception is thrown at Cluster#onUp:1496.  f.get() throws an ExecutionException, which bypasses updateCreatedPools (I don't suspect that is a problem but maybe it is):

```
179040 [cluster1-reconnection-1] ERROR com.datastax.driver.core.Cluster  - Unexpected error while setting node up
java.util.concurrent.ExecutionException: com.datastax.driver.core.OperationTimedOutException: [/127.0.1.3:9042] Operation timed out
    at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:306)
    at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:293)
    at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
    at com.datastax.driver.core.Cluster$Manager.onUp(Cluster.java:1496)
    at com.datastax.driver.core.Cluster$Manager.onUp(Cluster.java:1417)
    at com.datastax.driver.core.Cluster$Manager.access$800(Cluster.java:1116)
    at com.datastax.driver.core.Cluster$Manager$5.onReconnection(Cluster.java:1667)
    at com.datastax.driver.core.AbstractReconnectionHandler.run(AbstractReconnectionHandler.java:124)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:439)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
    at java.util.concurrent.FutureTask.run(FutureTask.java:138)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
    at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
    at java.lang.Thread.run(Thread.java:695)
Caused by: com.datastax.driver.core.OperationTimedOutException: [/127.0.1.3:9042] Operation timed out
    at com.datastax.driver.core.Connection$Future.onTimeout(Connection.java:1167)
    at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1230)
    at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)
    at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:655)
    at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)
    ... 1 more
```

After this it appears that no more attempts are made to connect to that host.  I produced this on the java505 branch running TimeoutStressTest, but I think I can produce it on this branch as well.   I'll try to see if I can understand this more on Monday.

EDIT: It seems this may not be the cause, since I've seen it happen twice for the same host within 20 seconds.  I do think there is a possibility that a host is given up on though, either that or TimeoutStressTest is too aggressive and can't establish connectivity to certain nodes after a while.  I'll continue investigating next week.

tolbertam committed:
Remove isClosing in pool creation failure.

tolbertam commented:
Removing the isClosing from the failure callback appears to have resolved the issue with a host not being retried.  After debugging I found that it could leave in a Pool in the state where it thinks it's closing, but is not.   Ran for an hour and there were 0 leaked connections.

tolbertam commented:
I've added some comments to #321, been  experimenting and what I am struggling with the most is handling the case where a non-initialized Pool is having borrowConnection invoked on it.   Shouldn't we not make uninitialized Pools available in SessionManager until they are initialized?  Previously a connection pool wasn't added to HostConnectionPool.pools until it had been initialized, but now that is not the case when replacing/renewing pools.  Would it be more ideal to have a separate map for uninitialized pools and the pool gets placed as soon as it completes initializing, or should we check for the pool initialization status in borrowConnection and throw an exception if not initialized yet?

The current code on this PR seems to be the best behaving.  Occasionally I see a leaking connection belonging to a HostConnectionPool that isn't part of SessionManager.pools.  After 1 hour of running TimeoutStressTest I typically observe 1-2 cases of this, but sometimes not.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/Connection.java:
```diff
@@ -674,12 +763,17 @@ public void shutdown() {
+                    return;
+                } else {
+                    eventLoop.execute(this);
+                }
```
This block can be simplified as:

```
if (eventLoop != null)
    eventLoop.execute(this);
```

Commit:
Do not add newly-created pools until they are fully initialized (JAVA-692)

adutra commented:
> Shouldn't we not make uninitialized Pools available in SessionManager until they are initialized? Previously a connection pool wasn't added to HostConnectionPool.pools until it had been initialized, but now that is not the case when replacing/renewing pools. 

I am proposing to wrap the calls to `SessionManager.pools.put()` with callbacks so that the pool is only added to the collection when it is fully initialized.

Can you check if you are still seeing leaking connections?

olim7t commented:
Regarding the leak mentioned on the other PR (a connection that is not closed and has poolRef set to a HostConnectionPool with an initFuture marked with an 'OperationTimedoutException'), I suspect the connection's initFuture gets cancelled but the connection is not properly closed.

Commit:
Do not trigger pool initialization inside HostConnectionPool's constructor (JAVA-692)

tolbertam committed:
Check initialization status in borrowConnection and addConnection.

tolbertam committed:
Revert forceRenewPool to set future to false on Pool init failure.

tolbertam committed:
Revert maybeAddPool to set future to false on Pool init failure.

olim7t committed:
Defunct connection if initialization fails.

olim7t committed:
Use sameThreadExecutor for initialization future transformations.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -86,38 +94,55 @@ public void run() {
+        this.initFuture = SettableFuture.create();
+    }
+
+    public void initAsync() {
```
A minor enhancement would be to make this method return initFuture (cf. similar methods such as closeAsync) and make initFuture field private.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/SessionManager.java:
```diff
@@ -220,68 +202,83 @@ ListeningExecutorService blockingExecutor() {
+                logger.error("Error creating pool to " + host, t);
+                future.set(false);
             }
         });
```
We are not using the blocking executor here, and the parameter `executor` could be removed.

olim7t replied:
Yes, doing it.

adutra commented on driver-core/src/main/java/com/datastax/driver/core/SessionManager.java:
```diff
@@ -294,34 +291,39 @@ private boolean replacePool(Host host, HostDistance distance, HostConnectionPool
-                    logger.error("Error creating pool to " + host, e);
-                    return false;
-                }
+                }, executor);
```
Here, contrary to forceRenewPool, we are using the blocking executor - is this per design? The task doesn't seem to require it.

olim7t replied:
Correct, the task is very short so it can use sameThreadExecutor.

adutra commented on driver-core/src/test/java/com/datastax/driver/core/TimeoutStressTest.java:
```diff
@@ -0,0 +1,261 @@
+                // Ensure that we don't exceed maximum connections.  Log as warning as there will be a bit of a timing
+                // factor between retrieving open connections and checking the reaper.
+                if(openChannels.size() > maximumExpected) {
+                    logger.warn("{} of open channels: {} exceeds maximum expected: {}: {}", openChannels.size(),
```
The number of expected arguments does not match the actual one.

tolbertam replied:
Thanks, fixed!

tolbertam committed:
Add callback to close connection to handle cancellation of initFuture.

Additionally refactor Connection such that the initialization logic is
now moved into a separate method, initAsync.  This is parallel to the
changes in HostConnectionPool.

Reintroduce isInitialized to connection which is set to true if
initialization is successful.  Previously a connection may mistakenly be
marked as suspected if initFuture was cancelled and then an error occurs
on the connection.

Check for Pool being closed when initFuture completes as found a case
where Session and its Pools are closed, but identified a Pool where it
had Connections that were never closed.

tolbertam committed:
Update TimeoutStressTest to use same partition.

Noticed that since the query being executed was a range query it would
always reuse the same host first.

olim7t committed:
Remove unused arguments in method that don't require an executor anymore.

olim7t committed:
General cleanup.

- organize imports
- formatting
- some renaming

olim7t committed:
Refactor callbacks in onAdd and onUp.

olim7t commented on driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:
```diff
@@ -86,37 +90,54 @@ public void run() {
+                    open.set(l.size());
+                    logger.trace("Created connection pool to host {}", host);
+                    phase = Phase.READY;
+                    initFuture.set(null);
```
There might still be a race if the pool was closed after the check. Will take a closer look at it.

olim7t committed:
Provide hook to customize executor.

tolbertam committed:
Update TimeoutStressTest to ensure pool grows back to core after test.

Adds additional validation to the test such that it ensures that once
timeouts are restored to reasonable values that the pool grows back to
an expected core state.

Reduce duration of test to 1 minute.  As a default this test should
serve more as a sanity check to ensure connections don't leak.  In local
testing about 4000 channels are created within 1 minute which should be
enough time to find an obvious leak.

Add test for verifying PoolingOptions#setInitializationExecutor.

olim7t committed:
Revisit races between init and close.

olim7t commented:
Closing this pull request in favor of #321 (which was started from the java692 branch).
